torch_np/_ndarray.py

```diff
@@ -343,67 +399,80 @@ def __setitem__(self, index, value):
+    if order != 'K':
         raise NotImplementedError
 
-    if isinstance(a, ndarray):
-        if dtype is not None and dtype != a.dtype:
-            a = a.astype(dtype)
-        return a
+    # a happy path
+    if isinstance(object, ndarray):
+        if copy is False and dtype is None and ndmin <= object.ndim:
+            return object
 
-    if isinstance(a, (list, tuple)):
-        # handle lists of ndarrays, [1, [2, 3], ndarray(4)] etc
+    # lists of ndarrays: [1, [2, 3], ndarray(4)] convert to lists of lists
+    if isinstance(object, (list, tuple)):   
         a1 = []
-        for elem in a:
+        for elem in object:
             if isinstance(elem, ndarray):
                 a1.append(elem.get().tolist())
             else:
                 a1.append(elem)
+        object = a1
+
+    # get the tensor from "object"
+    if isinstance(object, ndarray):
+        tensor = object._tensor
+        base = object
+    elif isinstance(object, torch.Tensor):
+        tensor = object
+        base = None
     else:
-        a1 = a
+        tensor = torch.as_tensor(object)
+        base = None
 
-    torch_dtype = _dtypes.torch_dtype_from(dtype)
+        # At this point, `tensor.dtype` is the pytorch default. Our default may
+        # differ, so need to typecast. However, we cannot just do `tensor.to`,
+        # because if our desired dtype is wider then pytorch's, `tensor`
+        # may have lost precision:
 
-    # This and array(...) are the only places which talk to ndarray directly.
-    # The rest goes through asarray (preferred) or array.
-    out = ndarray()
-    tt = torch.as_tensor(a1, dtype=torch_dtype)
-    out._tensor = tt
-    return out
+        # int(torch.as_tensor(1e12)) - 1e12 equals -4096 (try it!)
 
+        # Therefore, we treat `tensor.dtype` as a hint, and convert the
+        # original object *again*, this time with an explicit dtype.
+        dtyp = _dtypes.dtype_from_torch(tensor.dtype)
+        default = _dtypes.get_default_dtype_for(dtyp)
+        torch_dtype = _dtypes.torch_dtype_from(default)
 
-def array(object, dtype=None, *, copy=True, order='K', subok=False, ndmin=0,
-          like=None):
-    _util.subok_not_ok(like, subok)
-    if order != 'K':
-        raise NotImplementedError
-
-    if isinstance(object, (list, tuple)):
-        obj = asarray(object)
-        return array(obj, dtype, copy=copy, order=order, subok=subok,
-                     ndmin=ndmin, like=like)
+        tensor = torch.as_tensor(object, dtype=torch_dtype)
 
-    if isinstance(object, ndarray):
-        result = object._tensor
-
-        if dtype != object.dtype:
-            torch_dtype = _dtypes.torch_dtype_from(dtype)
-            result = result.to(torch_dtype)
-    else:
+    # type cast if requested
+    if dtype is not None:
         torch_dtype = _dtypes.torch_dtype_from(dtype)
-        result = torch.as_tensor(object, dtype=torch_dtype)
+        tensor = tensor.to(torch_dtype)
+        base = None
```

Comment:
Note. This is alright because casting functions will be fused with creation functions, but in PyTorch eager would be better to simply use this dtype in the `as_tensor`s calls above.

Reply:
This circles back to an earlier discussion of who's the main intended consumer. 

E.g., emulation of `out=` arguments is certainly suboptimal for pytorch eager, as pytorch functions never receive the `out` tensors. It does make life easier for the wrapper though. 

So what do we aim to, and what is the balance between efficiency and simplicity we're aiming for.

Reply:
It's probably fine as it is. Let's optimise for a clean codebase for now.